nasa/run.sh

```diff
@@ -1,30 +1,41 @@
-#!/bin/bash
-# This is intended for running DPS jobs.
-
-# The environment.yaml file sets up a custom conda environment called `dem`.
-# For NASA MAAP DPS, use `source activate <custom>`, not `conda activate <custom>`
-source activate dem
+#!/usr/bin/env bash
 
-# [left  bottom  right top]
-INPUT_LEFT=$1
-INPUT_BOTTOM=$2
-INPUT_RIGHT=$3
-INPUT_TOP=$4
-COMPUTE=$5
+set -xeou pipefail
 
-# Get path to this run.sh script
-basedir=$( cd "$(dirname "$0")" ; pwd -P )
-REPO_ROOT_PATH=$(dirname ${basedir})
+# This is intended for running DPS jobs.
 
-# Per NASA MAAP DPS convention, all outputs MUST be placed
-# by the algorithm into a directory called "output".
-# Once the DPS job finishes, MAAP will copy everything from "output"
-# to a directory in my-public-bucket. Everything else on the instance
-# will be destroyed.
-mkdir -p ${PWD}/output
+# `bbox` is expected to be 4 space-separated coords: left bottom right top
+bbox=$1
+# `compute` is optional, but any non-empty value will be converted to `--compute`
+compute=${2:+--compute}
 
-# Setup the environment variables. (Req'd for sardem)
-export HOME=/home/ops
+# path to the base of this repo (one level up from this script)
+basedir=$(dirname "$(dirname "$(readlink -f "$0")")")
 
-python ${REPO_ROOT_PATH}/get_dem.py --bbox ${INPUT_LEFT} ${INPUT_BOTTOM} ${INPUT_RIGHT} ${INPUT_TOP} ${COMPUTE} --out_dir ${PWD}/output
+# Per NASA MAAP DPS convention, all outputs MUST be written under a directory
+# named 'output' relative to the current working directory.  Once the DPS job
+# finishes, MAAP will copy everything from 'output' to a directory under
+# 'my-private-bucket/dps_output'. Everything else on the instance will be lost.
+outdir="${PWD}/output"
 
+# - `sardem` requires `HOME` to be set, so if it is not set, set it to `/home/ops`.
+# - The environment.yaml file sets up a custom conda environment called `dem`,
+#   so we use `conda run` to run the script in that environment.
+# - Use `scalene` to profile the script and write the output to `profile.json`.
+# - The `---` is a separator to pass arguments to `get_dem.py`, not to `scalene`.
+# - Disable ShellCheck SC2086 because we want to pass the arguments unquoted to
+#   correctly specify inputs to `get_dem.py`, but ShellCheck will flag unquoted
+#   substitutions as problematic.  For example, we want to pass arguments like
+#   this:
+#
+#     get_dem.py --bbox -156 18.8 -154.7 20.3 --out_dir /path/to/output
+#
+#   and not like this:
+#
+#     get_dem.py --bbox "-156 18.8 -154.7 20.3" --out_dir /path/to/output
+#
+# shellcheck disable=SC2086
+AWS_NO_SIGN_REQUEST=YES HOME=${HOME:-/home/ops} \
+    "${CONDA_EXE:-conda}" run --live-stream --name dem \
+    scalene --json --outfile "${outdir}/profile.json" --- \
+    "${basedir}/get_dem.py" ${compute} --bbox ${bbox} --out_dir "${outdir}"
```

Comment:
It looks like ESA MAAP is using the python command directly ([link](https://github.com/MAAP-Project/get-dem/blob/3d42aba3fe93787528a8abe6a6389be215babf72/esa/build/entrypoint.sh#L27)) .

It might be better to activate the conda environment and then keep the python command similar to how it was previously, and not run via conda. This looks like the way to call scalene using that method: `python3 -m scalene [options] yourprogram.py` ([link to example in docs](https://github.com/plasma-umass/scalene/blob/master/docs/scalene-demo.ipynb))

Reply:
As per the other comment regarding `conda run` that @sujen1412 replied to, we're moving towards using `conda run` instead of explicit environment activation.

There are multiple ways to invoke `scalene`. I chose to directly invoke scalene rather than invoke it as a module, but there should be no difference in behavior. I just find it to be a bit more convenient.

In the notebook you linked, scalene's help message includes this bit:

```plain
command-line:
   % scalene [options] yourprogram.py
or
   % python3 -m scalene [options] yourprogram.py
```

I chose the former rather than the latter. However, if you prefer to use the latter, so that it is more similar in structure to how ESA is invoking the program, I can do that. That would make things look like so:

```plain
AWS_NO_SIGN_REQUEST=YES HOME=${HOME:-/home/ops} \
    "${CONDA_EXE:-conda}" run --live-stream --name dem \
    python3 -m scalene --json --outfile "${outdir}/profile.json" --- \
    "${basedir}/get_dem.py" ${compute} --bbox ${bbox} --out_dir "${outdir}"
```

Would you prefer that?

Reply:
Sweet!

@arthurduf , which method is preferred for ESA?

From my perspective, it'd be ideal if the command line calls for NASA and ESA are the same (or very similar).

Reply:
@nemo794 : I prefer the "python module" option , because we do not use a conda environment to package the get-dem. 

I would also say that it is better to have the same "command line call" but honestly I know what are the relevant difference between the "python module option" and the "conda run option"

Reply:
Since NASA uses `conda` and ESA does not, that will remain a difference between our respective `run.sh` scripts. Whether or not we directly invoke `scalene` or reference it as a module (`python3 -m scalene`) is a separate matter. I will change to the module usage per Sam's preference, but NASA will still use `conda run` as a prefix. ESA will simply not use the `conda run` prefix.